### DISCUSSION TITLE
Direct Memory Access

### DISCUSSION THREAD

**Discussion by kamikyo** [dotnet/runtime#53474] [Ideas] [answered]
Does .NET Core have the technology similar to DMA(Direct Memory Access) in java,?
Like `java.nio.channels.FileChannel.transgerTo` in java?

**Comment by huoyaoyuan** [comment, marked as answer]
DMA is a hardware technology and it depends on underlying OS to utilize it. For example, Socket.SendFile will let the OS to do the optimal way.

**Reply by kamikyo** [reply]
java is utilizing this hardware technology with this piece of **native code**, just like this:
**openjdk\jdk\src\solaris\native\sun\nio\ch\FileChannelImpl.c**
```
JNIEXPORT jlong JNICALL
Java_sun_nio_ch_FileChannelImpl_transferTo0(JNIEnv *env, jobject this,
                                            jint srcFD,
                                            jlong position, jlong count,
                                            jint dstFD)
{
#if defined(__linux__)
    off64_t offset = (off64_t)position;
    jlong n = sendfile64(dstFD, srcFD, &offset, (size_t)count);
    if (n < 0) {
        if (errno == EAGAIN)
            return IOS_UNAVAILABLE;
        if ((errno == EINVAL) && ((ssize_t)count >= 0))
            return IOS_UNSUPPORTED_CASE;
        if (errno == EINTR) {
            return IOS_INTERRUPTED;
        }
        JNU_ThrowIOExceptionWithLastError(env, "Transfer failed");
        return IOS_THROWN;
    }
    return n;
    ......
```

Do you mean that c# has similar code? If so that would be great! @huoyaoyuan

**Reply by shahid-pk** [reply]
this is interop with native code. Yes interop with native code is available in c#.

**Reply by huoyaoyuan** [reply]
See #46848

**Reply by InCerryGit** [reply]
> See #46848

This API doesn't seem to support `offset` and `count` parameters, is there no way to achieve file slice transfer?

**Comment by kamikyo** [comment]
net core always brings me surprises, thanks for your answer
